sparx/bin/sxctf_refine.py

```diff
@@ -0,0 +1,890 @@
+    def_step_size,
+):
+    """
+    Runs defocus refinemenet for a single particle. It will create a reprojection of the particle
+    given the known projection angle and shift parameters. It will use this reprojection to
+    create multiple version of the projection each filtered with a different ctf. The CTFs used have
+    different defoci in range given by the current defocus +- defocus search range.
+
+    :param particle_volume: Particle volume
+    :param particle_image: 2D particle image
+    :param projection_angle_and_shift: Projection parameters
+    :param current_ctf: Estimated CTF for the 2D particle image (micrograph)
+    :param def_search_range: Defocus search range in micrometer
+    :param def_step_size: Defocus step size in micrometer
+    :return: CTF with highest correlation, the estimated
+    error (standard deviation) and significance value
+    """
+
+    # #########
+    # Calculate projection
+    # #########
+    projection_2d = calc_2d_projection(
+        particle_volume, projection_angle_and_shift, return_real=False
+    )
+
+    # #########
+    # Calculate particle mask
+    # #########
+    mask2d = np.copy(projection_2d.do_ift().get_2dview())
+
+    mask2d = ndimage.gaussian_filter(mask2d, 3)
+    object_pixel = np.where(mask2d > 0.05)
+
+    ctf_list = create_ctf_list(current_ctf, def_search_range, def_step_size)
+
+    # #########
+    # Calculate all masks for bootstrapping
+    # #########
+    num_bootstrap_runs = 15
+    masks = [
+        create_mask(particle_image.get_2dview().shape, object_pixel, fraction_size=0.25)
+        for _ in range(num_bootstrap_runs)
+    ]
+    masks.append(
+        create_mask(particle_image.get_2dview().shape, object_pixel, fraction_size=1.0)
+    )
+
+    # #########
+    # Calculate CTF corrupted projections
+    # #########
+    projections_ctf_applied_fft = [apply_ctf(projection_2d, ctf) for ctf in ctf_list]
+    projections_ctf = [proj.do_ift() for proj in projections_ctf_applied_fft]
+
+    # #########
+    # Calculate the best ctf for each mask
+    # #########
+    # TODO: Instead of brute force, use smarter optimization
+    index_best_projections = [
+        find_index_best_projection(particle_image, projections_ctf, mask)
+        for mask in masks
+    ]
+    best_ctf_no_boostrap = ctf_list[index_best_projections[-1]]
+
+    # #########
+    # Estimate error and difference/error ratio
+    # #########
+    best_ctfs_bootstrap = [ctf_list[i] for i in index_best_projections[:-1]]
+    defocus_bootstrap = [ctf.defocus for ctf in best_ctfs_bootstrap]
+    error_sd = np.std(defocus_bootstrap)
+    drratio = np.abs(best_ctf_no_boostrap.defocus - current_ctf.defocus) / (
+        error_sd + 0.0005
+    )
+
+    # #########
+    # Cleaning!
+    # #########
+    del projections_ctf
+    del projection_2d
+    del ctf_list
+
+    return best_ctf_no_boostrap, error_sd, drratio
+
+
+# Python2 multiprocessing trick. Should replaced with change to python 3.6.
```

Comment:
Thought about it.... Maybe we should make it MPI in the first place

Reply:
I would not put to mush effort in this for the 1.2 release. I will do small changes in the optimization algorithm, which already give you a speed up of multiple times. However, if the changes for MPI are minor, I can do it.